sourcegraph/sourcegraph
# gitserver: make vcs syncer implementation request a fresh clone URL in between each remote git operation

Fixes: https://github.com/sourcegraph/sourcegraph/issues/60682 

The key insight into fixing https://github.com/sourcegraph/sourcegraph/issues/60682 is to make our `vcssyncer` implementation (that performs the fetch and clone operations) stop using a "static" remote URL for all of it's operations. 

Before, when we used a Github App installation, Sourcegraph will receive an [authentication token](https://docs.github.com/en/rest/apps/installations?apiVersion=2022-11-28#revoke-an-installation-access-token) that's valid for one hour. We embed this token in the remote URL that we use to configure git for the fetch and git remote show head (to figure out what the default branch is pointing to) operations. When a huge repository like https://github.com/sgtest/megarepo gets cloned - it's likely that it will take > 1 hour to clone. This means that the auth information embedded the remote URL will expire in between the fetch and remote head operations - which would fail the overall "clone" operation. 

The solution to this is to introduce a new  abstraction: 

```go
// RemoteURLSource is a source of a remote URL for a repository.
//
// The remote URL may change over time, so it's important to use this interface
// to get the remote URL every time instead of caching it at the call site.
type RemoteURLSource interface {
	// RemoteURL returns the latest remote URL for a repository.
	RemoteURL(ctx context.Context) (*vcs.URL, error)
}
```

The `RemoteURLSource` encapsulates all the logic to figure out what remote URL to use for a given repository (including authentication information, token refreshing, etc.). Instead of passing around a static remote URL, the `RemoteURLSource` is stored as a struct member on each of the syncers. Whenever the syncer performs a git operation (like fetch) - it consults the source for the latest remoteURL to use - and only uses that url for the current operation. This means that individual git operations now have the chance to "refresh" the token in the remote URL (if necessary) which solves the original problem.


---

Test plan:

- Create a github app installation using https://sourcegraph.com/docs/admin/external_service/github#using-a-github-app
- Add a humongous repository such as https://github.com/sgtest/megarepo to the instance that will take over an hour to clone
- See that the clone actually completes successfully
